### PR TITLE
intermediate.2017Q1.Russ.Savage

### PR DESCRIPTION
**Pull request by RX7RUSS**
# Checkpoint Rubric

This is the rubric that your instructor will use to grade your checkpoints. Please do not edit. 

## Checkpoint 1: Towers of Hanoi Terminal App
- [x] All tests passed: 55 points
- [x] Properly commented code: 15 points
- [x] Properly indented code: 15 points
- [x] Demonstrated effective use of JavaScript: 15 points

Specific Comments:

*Comments may also be made on the file itself using github's commenting tools.*

## Checkpoint 2: Towers of Hanoi GUI App
- [x] All tests passed: 55 points
- [x] Properly commented code: 15 points
- [x] Properly indented code: 15 points
- [x] Demonstrated effective use of jQuery: 15 points

Specific Comments:

*Comments may also be made on the file itself using github's commenting tools.*

## Checkpoint 3: Final Project Gist Blog
- [ ] Effective use of AJAX: 20 points
- [ ] Effective use of loops: 20 points
- [ ] Effective use of github API: 20 points
- [ ] Style and design: 10 points
- [ ] Properly commented code: 15 points
- [ ] Properly indented code: 15 points

Specific Comments:

*Comments may also be made on the file itself using github's commenting tools.*

### CODE REVIEW

File: apps/02RockPaperScissors.js [hunk 1]
```diff
@@ -7,10 +7,43 @@ var rl = readline.createInterface({
...
+   if (hand2.toLowerCase() === 'paper') {
+return 'Hand two wins!';}
+}
+
+if (hand1.toLowerCase() === 'paper') {
+  if (hand2.toLowerCase() === 'scissors') {
+    return 'Hand two wins!';
+  }
+   if (hand2.toLowerCase() === 'rock'){
+  return 'Hand one wins!';
```
Comment (rbmcintosh) [inline apps/02RockPaperScissors.js]: indent 2 spaces here.

File: apps/02RockPaperScissors.js [hunk 2]
```diff
@@ -7,10 +7,43 @@ var rl = readline.createInterface({
...
+   }
+}
+
+if (hand1.toLowerCase() === 'scissors') {
+  if (hand2.toLowerCase() === 'rock') {
+    return 'Hand two wins!';
+  }
+   if (hand2.toLowerCase() === 'paper'){
+  return 'Hand one wins!';
```
Comment (rbmcintosh) [inline apps/02RockPaperScissors.js]: indent 2 spaces here.

File: apps/02RockPaperScissors.js [hunk 3]
```diff
@@ -7,10 +7,43 @@ var rl = readline.createInterface({
...
+  if (hand1.toLowerCase() === hand2.toLowerCase()) {
+  return "It's a tie!";
+}
+
+if (hand1.toLowerCase() === 'rock') {
+  if (hand2.toLowerCase() === 'scissors') {
+    return 'Hand one wins!';
+  }
+   if (hand2.toLowerCase() === 'paper') {
+return 'Hand two wins!';}
```
Comment (rbmcintosh) [inline apps/02RockPaperScissors.js]: indent 2 spaces from the "if" before.

File: apps/02RockPaperScissors.js [hunk 4]
```diff
@@ -7,10 +7,43 @@ var rl = readline.createInterface({
...
 
+  // Hand one is just a variable that will just have the word Rock Paper or Scissor in it!
+
+  //Spec 1 - If it's a tie, return "It's a tie!" It must be exactly "It's a tie!" for the tests to pass
+  // Write code here
+
 
 function rockPaperScissors(hand1, hand2) {
+  if (hand1.toLowerCase() === hand2.toLowerCase()) {
+  return "It's a tie!";
```
Comment (rbmcintosh) [inline apps/02RockPaperScissors.js]: indent 2 spaces here.

File: apps/01PigLatin.js [hunk 5]
```diff
@@ -7,12 +7,55 @@ var rl = readline.createInterface({
...
+  
+  
+  var firstPart = word.slice(0, vowelIndex);
+  var restWord  = word.slice(vowelIndex, word.length)
+
+//Conditional added to notify the user if anything other than a word is input.
+
+  if(vowelIndex > 0) {
+    return restWord + firstPart + 'ay';
+  } else if (vowelIndex === 0) {
```
Comment (JamesPuckett) [inline apps/01PigLatin.js:52]: This solution for returning words beginning with vowels is great and much more simple than the way I did it. Great work.

File: apps/02RockPaperScissors.js [hunk 6]
```diff
@@ -7,10 +7,43 @@ var rl = readline.createInterface({
   output: process.stdout
 });
 
+  // Hand one is a variable that will just have the word Rock Paper or Scissor in it!
+
+  //Spec 1 - If it's a tie, return "It's a tie!" It must be exactly "It's a tie!" for the tests to pass
+
 
 function rockPaperScissors(hand1, hand2) {
```
Comment (JamesPuckett) [inline apps/02RockPaperScissors.js:15]: At the top of your PigLatin app, you took word and immediately turned it to word.toLowerCase(), I'm curious why you didn't do that with hand1 and hand2 here.

File: apps/02RockPaperScissors.js [hunk 7]
```diff
@@ -7,10 +7,43 @@ var rl = readline.createInterface({
...
+  if (hand1.toLowerCase() === hand2.toLowerCase()) {
```
Comment (JamesPuckett) [inline apps/02RockPaperScissors.js:16]: For this area I'm curious as to what would happen if you typed in two things that weren't rock, paper, or scissors? Like 'apple' & 'apple'?

File: apps/02RockPaperScissors.js [hunk 8]
```diff
@@ -7,10 +7,43 @@ var rl = readline.createInterface({
...
+    return "It's a tie!";
+  }
+
+  if (hand1.toLowerCase() === 'rock') {
```
Comment (JamesPuckett) [inline apps/02RockPaperScissors.js:20]: All of your if/else if/else statements are written and indented well make it very easy to read and understand. Nice work!

File: apps/03TicTacToe.js [hunk 9]
```diff
@@ -14,34 +14,70 @@ var board = [
...
-  // Your code here
-}
+  if((board[0][0] === playerTurn && board[1][1] === playerTurn && board[2][2] === playerTurn) ||
+  (board[0][2] === playerTurn && board[1][1] === playerTurn && board[2][0] === playerTurn)) {
+    return true;
+  };
+};
+
+// New game function will reset the board variable after any of the wins return true.
+function newGame() {
```
Comment (JamesPuckett) [inline apps/03TicTacToe.js]: I really like how you designed this function and called it in your TTT(). It makes it very easy to read and understand.

File: apps/02RockPaperScissors.js [hunk 10]
```diff
@@ -7,10 +7,43 @@ var rl = readline.createInterface({
...
+    if (hand2.toLowerCase() === 'rock') {
+      return 'Hand one wins!';
+    }
+  }
+
+  if (hand1.toLowerCase() === 'scissors') {
+    if (hand2.toLowerCase() === 'rock') {
+      return 'Hand two wins!';
+    }
+    if (hand2.toLowerCase() === 'paper') {
```
Comment (pieroapretto) [inline apps/02RockPaperScissors.js:42]: Nice job scrubbing your data but you should call the `.toLowerCase()` method only once for each parameter.

You can do so by storing the result of the method to a variable and then comparing it to 'rock', 'scissors', ectc.

File: apps/04TowersOfHanoi.js [hunk 11]
```diff
@@ -13,30 +13,78 @@ var stacks = {
...
 
+function isLegal(startStack, endStack) {
+  // The 2 variables commented out below are not necessary in this method. They are from a previous approach I tried that didn't work. 
+  //   var start = stacks[startStack];
+  //   var end = stacks[endStack];
+
+  if (isEmpty(startStack, endStack)) {
+    return true;
+  }
+  if (stacks[startStack].slice(-1) < stacks[endStack].slice(-1)) {
```
Comment (pieroapretto) [inline apps/04TowersOfHanoi.js]: impressive use of the `slice()` method!

File: apps/04TowersOfHanoi.js [hunk 12]
```diff
@@ -13,30 +13,78 @@ var stacks = {
...
-  // Your code here
-
+  // I used the .toLowerCase method to scrub the input data like we did in Pig Latin.
+
+  startStack = startStack.toLowerCase();
+  endStack = endStack.toLowerCase();
+  
+  if (isLegal(startStack, endStack)) {
+    movePiece(startStack, endStack);
+    checkForWin();
```
Comment (pieroapretto) [inline apps/04TowersOfHanoi.js:82]: Clean logic!

File: homework/01LessonOne.js [hunk 13]
```diff
@@ -17,27 +17,27 @@ var myUndefined;
...
 // Let's create a new variable named myNumber. Assign it a number.
 
-var myNumber;
+var myNumber = 94;
 
 // Problem 5:
 // Let's create a new variable named myString. Assign it a string.
 
-var myString;
+var myString = "Russ is cool";
```
Comment (pieroapretto) [inline homework/01LessonOne.js:40]: Russ is very cool!

File: homework/01LessonOne.js [hunk 14]
```diff
@@ -106,7 +106,7 @@ var notNotTrue;
 // What is the difference between the "and" and the "or" operators? Explain why you would use each of them.
 //
 // Your Answer Goes Here:
-//
+// The "and" operator checks if both the statements are true, and the "or" operator checks if one of the statements is true.
```
Comment (pieroapretto) [inline homework/01LessonOne.js:109]: Correct.

File: homework/03LessonThree.js [hunk 15]
```diff
@@ -89,22 +93,22 @@ var problemSevenAnswer;
...
+var emptyArray = [];
 
 // Problem 9:
 // Create an array with 5 elements in it
-var lengthFiveArray;
+var lengthFiveArray = [ 1, 2, 3, 4, 5 ];
 
 // DO NOT MODIFY
-var nameArray = ['bob', 'fred', 'susan'];
+var nameArray = ['bob', 'george', 'susan'];
```
Comment (pieroapretto) [inline homework/03LessonThree.js:103]: Do not modify threeByThreeArrray or nameArray directly.

You need to use bracket or dot notation to access and assign values to these arrays.

File: homework/04LessonFour.js [hunk 16]
```diff
@@ -76,6 +92,8 @@ var nestedObjects = {
     }
   }
 };
+
+nestedObjects['users']['bob roberts']['starring'] = 'Tim Robbins';
```
Comment (pieroapretto) [inline homework/04LessonFour.js:96]: Great use of bracket notation here!

dancingmarshmallow

File: apps/05Mastermind.js [hunk 17]
```diff
@@ -29,12 +29,47 @@ function getRandomInt(min, max) {
...
+  if (guess === solution) {
+    return 'You guessed it!';
+  }
+  if (board.length === 10) {
+    return ('You ran out of turns! The solution was ' + solution);
+  }
+  else {
+    var hint = generateHint(solution, guess);
+    board.push(hint +' '+ guess);
+    return ('Guess again');
```
Comment (pieroapretto) [inline apps/05Mastermind.js:71]: All tests passed!

File: apps/06.1Checkers.js [hunk 18]
```diff
@@ -52,15 +77,82 @@ function Board() {
...
-    // Your code here
+    this.board.createCheckers();
+  };
+
+  this.moveChecker = function(start, end) {
+    var checker = this.board.selectChecker(start[0], start[1]);
+    this.board.grid[start[0]][start[1]] = null;
+    this.board.grid[end[0]][end[1]] = checker;
+    if (Math.abs(start[0] - end[0]) === 2) {
+      var killPosition = [(parseInt(start[0],10) + parseInt(end[0],10)) / 2 , (parseInt(start[1],10) + parseInt(end[1],10)) / 2];
```
Comment (pieroapretto) [inline apps/06.1Checkers.js:153]: Nice variable logic.

You can simplify this logic further by calling the parseInt() methods on the start and end arguments directly. JavaScript treats simple strings as array types.

`var midpoint = ((parseInt(end) + parseInt(start)) / 2).toString();`

File: apps/06.2SpaceTravelWithComments.js [hunk 19]
```diff
@@ -0,0 +1,235 @@
...
+    this.crew = [];
+
+    //again, the missionStatement function should only work with ships, so make it a method
+    this.missionStatement = function(){ //where did the ship parameter go?
+      //remember.. what is in the "this" variable at this point
+        //lets use ES6 for/of
+        //what does of do 
+        //it simplifies a wordy pattern
+        //get rid of a wordy for loop and we didn't have to get the index of the specific item in the array
+      for(var crewmember of this.crew){  // for(var i =0; i < myship.crew.length;i++)
```
Comment (pieroapretto) [inline apps/06.2SpaceTravelWithComments.js:143]: Nice for loop!

File: apps/06.1Checkers.js [hunk 20]
```diff
@@ -52,15 +77,82 @@ function Board() {
...
+      this.board.killChecker(killPosition);
```
Comment (pieroapretto) [inline apps/06.1Checkers.js:154]: All test passed! Neato!

wizard

File: apps/07jquery-tic-tac-toe/script.js [hunk 21]
```diff
@@ -1,5 +1,69 @@
...
   // Put app logic in here
+  var playerTurn = 'X';
+  
+  $('[data-cell]').click(function () {
+    $(this).text(playerTurn);
+    checkForWin();
+    playerTurn = (playerTurn === 'X') ? 'O' : 'X';
+    newGame();
+    
+  })
```
Comment (pieroapretto) [inline apps/07jquery-tic-tac-toe/script.js:11]: missing semicolon.

File: apps/08jquery-towers-of-hanoi/script.js [hunk 22]
```diff
@@ -1,5 +1,51 @@
...
+      if($blockToBePlaced < $lastBlockValue || !$lastBlockValue) {
+        //comparing the block that was clicked on to the last block in a stack
+        
+        $(this).append($block);
+        $block = null;
+        checkForWin();
+      }
+              
+    } else {
+      $block = $(this).children().last().detach();
```
Comment (pieroapretto) [inline apps/08jquery-towers-of-hanoi/script.js:29]: If you make `$lastBlock accessible to the scope of the whole function you can avoid duplicate logic here.

Ex:
`$block = $lastBlock.detach();`

File: apps/08jquery-towers-of-hanoi/script.js [hunk 23]
```diff
@@ -1,5 +1,51 @@
...
+    // $block = $(this).children().last().detach();
+      // The above variable and methods were moved to the else statement below
+    
+    if($block) {
+      var $lastBlock = $(this).children().last();
+      var $lastBlockValue = $lastBlock.data("block");
+      var $blockToBePlaced = $block.data("block");
+
+
+      if($blockToBePlaced < $lastBlockValue || !$lastBlockValue) {
```
Comment (pieroapretto) [inline apps/08jquery-towers-of-hanoi/script.js:20]: Nice shorthand code!

File: homework/08LessonEight/script.js [hunk 24]
```diff
@@ -17,30 +17,30 @@ $(document).on('ready', function() {
...
   // use $finalList.prepend() to put window.firstListChildren
   // at the beggining of #final-list
-  $finalList;
+  $finalList.prepend(window.firstListChildren);
 
   // Problem 6:
   // use $finalList.append() to put window.secondListChildren
   // at the end of #final-list
-  $finalList;
+  $finalList.append(window.secondListChildren);
```
Comment (pieroapretto) [inline homework/08LessonEight/script.js:44]: All tests passed. Very nice work.

File: homework/07LessonSeven/script.js [hunk 25]
```diff
@@ -69,20 +69,22 @@ $(document).on('ready', function() {
...
   // and set window.dataIdValue to the value of the data-id attribute of the element
   // you can do this by calling .data() on the selected element
-  window.dataIdValue;
+  
+  window.dataIdValue = $('#data-element').data('id');
 
   // Problem 12:
   // select the element with id="data-element"
   // and set its text (.text()) to the value of the data-content attribute
+  $('#data-element').text($('#data-element').attr('data-content'));
```
Comment (pieroapretto) [inline homework/07LessonSeven/script.js:89]: Nicely done.

File: apps/10address-book-ajax/script.js [hunk 26]
```diff
@@ -14,22 +12,24 @@ $(document).ready(function () {
...
                 <p>${user.occupation}</p>
+                <p>${user.phone}</p>
                 <p>${user.address}</p>
                 <img src='${user.avatar}'>`;
-              console.log(str2);
-              $('#details').html(str2);
-            }
-          })
+            console.log(str2);
+            $('#details').html(str2);
```
Comment (pieroapretto) [inline apps/10address-book-ajax/script.js]: Very neato work.

File: apps/11gist-blog/script.js [hunk 27]
```diff
@@ -1,5 +1,57 @@
...
+    }
+    $.ajax(commentURL, getComments);
+  }
+};
+
+
 $(document).ready(function() {
-  // You code here
+  // first ajax call
+  $.ajax("http://127.0.0.1:8080/apps/11gist-blog/api/gists.json", gistDescription)
```
Comment (pieroapretto) [inline apps/11gist-blog/script.js:50]: You can shorten this to `$.ajax("api/gists.json", gistDescription)` since the path is relative. Similar to how you would route css files in a html page.

File: apps/11gist-blog/script.js [hunk 28]
```diff
@@ -1,5 +1,57 @@
...
+// callback to get the content of the gist that was clicked
+var gistContent = {
+  success: function(gist) {
+    var content = gist.files['post.md']['content'];
+    var commentURL = gist["comments_url"];
+    var str = 
+    ('<div id="post">' +
+    '<p>' + marked(content) + '</p>' + '</div>')
+    $('#post').replaceWith(str);
+    var getComments = {
```
Comment (pieroapretto) [inline apps/11gist-blog/script.js:29]: Please nest your `getComments` callback outside of the `gistContent` callback like you handle `gistContent` on line 55. 

This will make your logic much easier to read.

File: apps/11gist-blog/script.js [hunk 29]
```diff
@@ -1,5 +1,57 @@
...
-  // You code here
+  // first ajax call
+  $.ajax("http://127.0.0.1:8080/apps/11gist-blog/api/gists.json", gistDescription)
+  $('body').on('click', 'a', function(event) {
+    event.preventDefault();
+    // variable to store the url in the data attribute
+    var gistURL = $(this).data('url');
+    $.ajax(gistURL, gistContent);
+  });
 });
```
Comment (pieroapretto) [inline apps/11gist-blog/script.js:57]: All "test" passed! Great work on your final.

grad
